### ISSUE TITLE
Necessity of multiple address ranges for an IP (memories and control registers in same addr space)

### ISSUE THREAD

**Issue by eunchan** [lowRISC/opentitan#2957]
Want to get the opinions from SW ppl about the case of OTP and any IPs could have memory-mapped SRAM,non-volatile memories with its control reigsters (e.g. OTP, flash_ctrl + flash_macro, SPI_DEVICE (DPSRAM + control))

@msfschaffner @tjaychen and I expect that having them in the same address space makes software ACL of the address space more difficult or requires fine-granularity of the PMP). 

If separated address space for control register is required, modifications below may needed:

1. `topgen` to have multiple address ranges in `module`, `memory` data structure.
2. IP modification
2-1. IP design and hjson to have mutliple bus device ports.
2-2. IP.hjson to specify how many address ranges its design has and create tl window based on it. (single TL-UL port)
3. `reggen` to get the base address of registers and parse only based on that and ignore other address. It needs to export the base address parameter to the top
4. `tlgen` modification
4-1 (multiple ports). `tlgen` to accept multiple TL-UL device port for an IP and assign the names for each TL port correctly.
4-2 (single port). Having multiple address ranges in a TL-UL port in an IP is supported in `tlgen`.

CC: @msfschaffner @sjgitty @tjaychen

**Comment by msfschaffner** [comment]
I believe the multiple TL-ports extension we added last year solved this issue?

CC @rswarbrick

**Comment by rswarbrick** [comment]
Yep reggen now allows an IP block to have two address ranges. *But* it does so by creating two TL devices, with the corresponding increase in crossbar area and doubled decoding logic.

What we don't currently support is "I've got one device, but it lives at two disjoint address ranges, with something else in between". My guess is that this isn't worth thinking too hard about, especially as I suspect you'd still have the same crossbar in either case. If we *really* wanted to support that, we could add a "coalesce my devices" option in the hjson and then spit out a single reg_top. But it's a bit awkward, because you'd have to pass through some extra information about base addresses from the top-level to handle everything properly.

**Comment by tjaychen** [comment]
i personally don't think that case is really worth supporting.
If someone does have two distinct address spaces like that... it's better to just model it the way you've shown with two separate reg blocks. 

are you guys okay with closing this?

**Comment by tjaychen** [comment]
lets do it! (i'll close it first :) )
